api/client/cp.go

```diff
@@ -10,48 +15,242 @@ import (
+}
+
+func resolveLocalPath(localPath string) (absPath string, err error) {
+	if absPath, err = filepath.Abs(localPath); err != nil {
+		return
+	}
+
+	return archive.PreserveTrailingDotOrSeparator(absPath, localPath), nil
+}
+
+func (cli *DockerCli) copyFromContainer(srcContainer, srcPath, dstPath string) (err error) {
+	if dstPath != "-" {
+		// Get an absolute destination path.
+		dstPath, err = resolveLocalPath(dstPath)
+		if err != nil {
+			return err
+		}
+	}
+
+	query := make(url.Values, 1)
+	query.Set("path", filepath.ToSlash(srcPath)) // Normalize the paths used in the API.
+
+	urlStr := fmt.Sprintf("/containers/%s/archive-path?%s", srcContainer, query.Encode())
+
+	response, hdr, statusCode, err := cli.call("GET", urlStr, nil, nil)
 	if err != nil {
 		return err
 	}
+	defer response.Close()
 
-	hostPath := cmd.Arg(1)
-	if statusCode == 200 {
-		if hostPath == "-" {
-			_, err = io.Copy(cli.out, stream)
-		} else {
-			err = archive.Untar(stream, hostPath, &archive.TarOptions{NoLchown: true})
+	if statusCode != 200 {
+		return fmt.Errorf("unexpected status code from daemon: %d", statusCode)
+	}
+
+	if dstPath == "-" {
+		// Send the response to STDOUT.
+		_, err = io.Copy(os.Stdout, response)
+
+		return err
+	}
+
+	var stat types.ContainerPathStat
+	encodedStat := hdr.Get("X-Docker-Container-Path-Stat")
+	statDecoder := base64.NewDecoder(base64.StdEncoding, strings.NewReader(encodedStat))
+	err = json.NewDecoder(statDecoder).Decode(&stat)
+	if err != nil {
+		return fmt.Errorf("unable to decode container path stat header: %s", err)
+	}
+
+	// Prepare source copy info.
+	srcInfo := archive.CopyInfo{
+		Path:   srcPath,
+		Exists: true,
+		IsDir:  stat.Mode.IsDir(),
+	}
+
+	return archive.CopyTo(response, srcInfo, dstPath)
+}
+
+func (cli *DockerCli) copyToContainer(srcPath, dstContainer, dstPath string) (err error) {
+	if srcPath != "-" {
+		// Get an absolute source path.
+		srcPath, err = resolveLocalPath(srcPath)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Prepare destination copy info by stat-ing the container path.
+	dstInfo := archive.CopyInfo{Path: dstPath}
+	dstStat, err := cli.statContainerPath(dstContainer, dstPath)
+	if err == nil {
+		dstInfo.Exists, dstInfo.IsDir = true, dstStat.Mode.IsDir()
+	}
+	// Ignore any other error and assume that the parent directory of the
+	// destination path exists, in which case the copy may still succeed. If
+	// there is any type of conflict (e.g., non-directory overwriting an
+	// existing directory or vice versia) the extraction will fail. If the
+	// destination simply did not exist, but the parent directory does, the
+	// extraction will still succeed.
+
+	var content io.Reader
+	if srcPath == "-" {
+		// Use STDIN.
+		content = os.Stdin
+		if !dstInfo.IsDir {
+			return fmt.Errorf("destination is not a directory")
+		}
+	} else {
+		srcArchive, err := archive.TarResource(srcPath)
+		if err != nil {
+			return err
 		}
+		defer srcArchive.Close()
+
+		// Prepare source copy info.
```

Comment:
Most of explanation for why this is done was left in the old PRs, an the exactly logic behind what 'PrepareArchiveCopy' does is documented in comments within that function itself, but I should add some comments here to tie it all together.

In the previous PRs leading up to this one, we came to the consensus that the API shouldn't be a "copy" API, but a more simple "archive" and "extract archive" API. "Copy" is a feature of the Docker CLI only which happens to use the archive/extract API.